File: VMSFrontend/src/app/core/trip.models.ts
/**
 * Trips — fixed and open, their lifecycle, list/search and history (FSD §21, §22, §24, §48.4 screens 12-15 +
 * the unnumbered "Trip List (desk)" row).
 */

export type TripType = 'Fixed' | 'Open';
export type TripStatus = 'Draft' | 'Planned' | 'Assigned' | 'Started' | 'InTransit' | 'AtPickup' | 'Loaded' | 'AtDelivery' | 'Delivered' | 'Completed' | 'OnHold' | 'Cancelled';
export type TripLocationType = 'City' | 'Other';
export type OtherLocationType = 'Warehouse' | 'Factory' | 'CustomerSite' | 'Depot' | 'Terminal' | 'ConstructionSite' | 'Other';
export type RateSource = 'Configured' | 'Manual' | 'Missing';

export interface LocationModel {
  locationType: TripLocationType;
  cityId?: number | null;
  otherLocationType?: OtherLocationType | null;
  otherLocationName?: string | null;
  otherNearestCityId?: number | null;
  label: string;
}

export interface LocationInput {
  locationType: TripLocationType;
  cityId?: number | null;
  otherLocationType?: OtherLocationType | null;
  otherLocationName?: string | null;
  otherNearestCityId?: number | null;
}

export interface TripModel {
  tripId: number;
  tripNumber: string;
  tripType: TripType;
  customerId: number;
  customerTripReference?: string | null;
  tripConfigurationId?: number | null;
  routeId?: number | null;
  vehicleId: number;
  driverId?: number | null;
  defaultDriverId?: number | null;
  isDriverOverridden: boolean;
  driverOverrideReason?: string | null;
  tripDate: string;
  plannedStart?: string | null;
  actualStart?: string | null;
  actualEnd?: string | null;
  startOdometer?: number | null;
  endOdometer?: number | null;
  tripRateId?: number | null;
  tripRateAmount?: number | null;
  rateEffectiveFrom?: string | null;
  rateEffectiveTo?: string | null;
  rateSource: RateSource;
  tripAmount?: number | null;
  currencyCode?: string | null;
  rateMissing: boolean;
  status: TripStatus;
  heldFromStatus?: string | null;
  holdReason?: string | null;
  cancelReason?: string | null;
  completionDate?: string | null;
  isActive: boolean;
  inactiveReason?: string | null;
  invoiceId?: number | null;
  remarks?: string | null;
  rowVersion: string;
  warnings: string[];
  // Open trips only
  from?: LocationModel | null;
  to?: LocationModel | null;
  stops: LocationModel[];
  isRoundTrip: boolean;
  routeLabel?: string | null;
}

export interface CreateFixedTripRequest {
  customerId: number;
  tripConfigurationId: number;
  vehicleId: number;
  driverId?: number | null;
  driverOverrideReason?: string | null;
  tripDate: string;
  customerTripReference?: string | null;
  plannedStart?: string | null;
  remarks?: string | null;
}

export interface CreateOpenTripRequest {
  customerId: number;
  vehicleId: number;
  driverId?: number | null;
  driverOverrideReason?: string | null;
  tripDate: string;
  customerTripReference?: string | null;
  plannedStart?: string | null;
  from: LocationInput;
  to: LocationInput;
  stops: LocationInput[];
  isRoundTrip: boolean;
  tripAmount: number;
  remarks?: string | null;
}

// ── Lifecycle (§24) ──────────────────────────────────────────────────────────────────

export interface TransitionTripRequest {
  startOdometer?: number | null;
  endOdometer?: number | null;
  rowVersion: string;
}

export interface HoldTripRequest {
  reason: string;
  rowVersion: string;
}

export interface ResumeTripRequest {
  rowVersion: string;
}

export interface CancelTripRequest {
  reason: string;
  rowVersion: string;
}

export interface ReopenTripRequest {
  rowVersion: string;
}

export interface ChangeTripActiveRequest {
  reason?: string | null;
  rowVersion: string;
}

// ── List / search (§48.4's unnumbered "Trip List (desk)" row) ──────────────────────────

export interface TripListItem {
  tripId: number;
  tripNumber: string;
  tripType: TripType;
  tripDate: string;
  customerId: number;
  customerName: string;
  customerTripReference?: string | null;
  routeLabel?: string | null;
  vehicleId: number;
  vehicleRegistrationNo?: string | null;
  driverId?: number | null;
  driverName?: string | null;
  status: TripStatus;
  tripAmount?: number | null;
  currencyCode?: string | null;
  rateMissing: boolean;
  podStatus?: string | null;
  invoiceId?: number | null;
  invoiceNumber?: string | null;
  isActive: boolean;
}

export interface TripSearchFilter {
  fromDate?: string | null;
  toDate?: string | null;
  customerId?: number | null;
  vehicleId?: number | null;
  driverId?: number | null;
  status?: string | null;
  tripType?: string | null;
  isActive?: boolean | null;
  invoiced?: boolean | null;
}

// ── History (§48.1) ──────────────────────────────────────────────────────────────────

export interface TripHistoryChange {
  id: number;
  occurredAt: string;
  groupId: string;
  userName?: string | null;
  entity: string;
  recordId: string;
  action: string;
  field?: string | null;
  oldValue?: string | null;
  newValue?: string | null;
  reason?: string | null;
  restricted: boolean;
}

export interface TripHistory {
  changes: { items: TripHistoryChange[]; totalCount: number; page: number; pageSize: number; totalPages: number };
}

File: src/VMS.Modules.Trips/Models/TripModels.cs
using VMS.Shared.Pagination;

namespace VMS.Modules.Trips.Models;

/// <summary>§48.1: "History button on every record" — one row per audited change to the trip itself or to any of
/// its child rows (events, stops, fuel, expenses, income, documents, POD, issues, rate history all root back to
/// "Trip" — see each entity's own <c>GetAuditRoot()</c>), newest first.</summary>
public sealed class TripHistoryChange
{
    public long Id { get; set; }
    public DateTime OccurredAt { get; set; }
    public Guid GroupId { get; set; }
    public string? UserName { get; set; }
    public string Entity { get; set; } = string.Empty;
    public string RecordId { get; set; } = string.Empty;
    public string Action { get; set; } = string.Empty;
    public string? Field { get; set; }
    public string? OldValue { get; set; }
    public string? NewValue { get; set; }
    public string? Reason { get; set; }
    public bool Restricted { get; set; }
}

public sealed class TripHistory
{
    public PaginatedResponse<TripHistoryChange> Changes { get; set; } = new();
}

public sealed class TripModel
{
    public long TripId { get; set; }
    public string TripNumber { get; set; } = string.Empty;
    public string TripType { get; set; } = string.Empty;
    public int CustomerId { get; set; }
    public string? CustomerTripReference { get; set; }
    public long? TripConfigurationId { get; set; }
    public int? RouteId { get; set; }
    public int VehicleId { get; set; }
    public int? DriverId { get; set; }
    public int? DefaultDriverId { get; set; }
    public bool IsDriverOverridden { get; set; }
    public string? DriverOverrideReason { get; set; }
    public DateOnly TripDate { get; set; }
    public DateTime? PlannedStart { get; set; }
    public DateTime? ActualStart { get; set; }
    public DateTime? ActualEnd { get; set; }
    public decimal? StartOdometer { get; set; }
    public decimal? EndOdometer { get; set; }
    public long? TripRateId { get; set; }
    public decimal? TripRateAmount { get; set; }
    public DateOnly? RateEffectiveFrom { get; set; }
    public DateOnly? RateEffectiveTo { get; set; }
    public string RateSource { get; set; } = string.Empty;
    public decimal? TripAmount { get; set; }
    public string? CurrencyCode { get; set; }
    public bool RateMissing { get; set; }
    public string Status { get; set; } = string.Empty;
    public string? HeldFromStatus { get; set; }
    public string? HoldReason { get; set; }
    public string? CancelReason { get; set; }
    public DateOnly? CompletionDate { get; set; }
    public bool IsActive { get; set; }
    public string? InactiveReason { get; set; }
    public long? InvoiceId { get; set; }
    public string? Remarks { get; set; }
    public string RowVersion { get; set; } = string.Empty;
    public IReadOnlyList<string> Warnings { get; set; } = [];

    // ── Open trips only (§22, CC-13) — null/empty for Fixed trips. ─────────────────────
    public LocationModel? From { get; set; }
    public LocationModel? To { get; set; }
    public IReadOnlyList<LocationModel> Stops { get; set; } = [];
    public bool IsRoundTrip { get; set; }
    /// <summary>§22: "the route label on invoices is built from stops, e.g. LHR → DGK" — computed here so
    /// whatever later reads a trip (an invoice line, a report) never re-derives it differently.</summary>
    public string? RouteLabel { get; set; }
}

public sealed class CreateFixedTripRequest
{
    public int CustomerId { get; set; }
    public long TripConfigurationId { get; set; }
    public int VehicleId { get; set; }
    /// <summary>Left blank to accept the vehicle's own default driver (§20); a value different from that default
    /// is an override, needing the TRP.TRIP.OVERRIDEDRIVER permission and a <see cref="DriverOverrideReason"/>.</summary>
    public int? DriverId { get; set; }
    public string? DriverOverrideReason { get; set; }
    public DateOnly TripDate { get; set; }
    public string? CustomerTripReference { get; set; }
    public DateTime? PlannedStart { get; set; }
    public string? Remarks { get; set; }
}
